### ISSUE TITLE
MemoryMappedViewAccessor: Dispose flushes on temporary files

### ISSUE THREAD

**Issue by msedi** [dotnet/runtime#94383]
**Edit by @adamsitnik:** the proposal: https://github.com/dotnet/runtime/issues/94383#issuecomment-2260137688

Since all my requests are on hold, I just wanted to open this issue again, coming from #59606 which I have closed but is still a problem that maybe needs further discussion. So for better readability I have copied the original issue description from the original post here and added a few finsings.

**Description**
Currently it seems that the MemoryMappedViewAccessor created from a MemoryMappedFile flushes when the view is disposed.
While this behavior is fully acceptable in regular scenarios where the View is created, written to and closed again. In this situation the modifications need to be flushed to disk.

Another possibility is though to use the MM-File as a temporary buffer, where the FileStream is created with FileOptions.DeleteOnClose and when the MM-FIle and the FileStream get closed, the temporary storage file is also deleted.

For my case I open many MM-Files (with the temporary flag), knowing the file is of all files in advance and setting the file size with SetLength (also using the WinAPI SetFileValidLength, and making it also a sparse file with winAPI IOControl). Then a full view is created.

So my question is if there is a solution to skip the flush (that I currently don't) or if I should make an API proposal that allows to skip the Dispose, or to advise the view to do a discard (by using DiscardVirtualMemory).

[Edit] It seems when disposing the MemoryMappedFile before the View works without flushing, but I don't know if this is the intended behavior? **Nevertheless when looking at my process with Process Monitor, I can still a write is performed.**

**Configuration**
My configuration currently is .NET 5/Win10/x64

**Regression**?
No, there is no regression. It is a problem that has existed for a longer time and maybe no one has ever encountered it since it has to to only with temporary files.

**Data**
An unfinished project can be seen here: https://github.com/msedi/MappableFileStream

**Comment by MichalPetryka** [comment]
> FileOptions.Temporary

You mean `FileOptions.DeleteOnClose`?

**Comment by msedi** [comment]
>  You mean `FileOptions.DeleteOnClose`?

Yes. Sorry I already clarified this in some other discussion, but I haven't corrected it here.

**Comment by adamsitnik** [comment]
> Another possibility is though to use the MM-File as a temporary buffer, where the FileStream is created with FileOptions.DeleteOnClose and when the MM-FIle and the FileStream get closed, the temporary storage file is also deleted.

Other .NET APIs like `BinaryReader` ctor already accept `bool leaveOpen` to prevent the stream from being closed on dispose. We could add one more overload with the new bool argument:

```diff
namespace System.IO.MemoryMappedFiles;

public partial class MemoryMappedFile : IDisposable
{
   public MemoryMappedViewAccessor CreateViewAccessor()
   public MemoryMappedViewAccessor CreateViewAccessor(long offset, long size)
   public MemoryMappedViewAccessor CreateViewAccessor(long offset, long size, MemoryMappedFileAccess access);
+  public MemoryMappedViewAccessor CreateViewAccessor(long offset, long size, MemoryMappedFileAccess access, bool leaveOpen);
}
```

> So my question is if there is a solution to skip the flush (that I currently don't)

Why do you want to skip the flush? It's not needed because the file won't be persisted?

**Comment by msedi** [comment]
> Other .NET APIs like `BinaryReader` ctor already accept `bool leaveOpen` to prevent the stream from being closed on dispose. We could add one more overload with the new bool argument:

Sorry for the delay. 

I'm using the memory-mapped file as a temporary buffer that is only needed as long the application runs or as long as the memory-mapped file exists. If the temporary buffer is disposed it should die immediately.

We has a similar issue with file-streams. The filestream is initialized with a fixed length, e.g. 60GB. Then the filestream is used as a backing-buffer (like the memory-mapped file). When the file is closed all data is written to disk, even the regions in the file that have not been touched. This means, that even if I write nothing into the file, the data is written. We got around this by setting the sparsity flag using `DeviceIoControl` but this is fortunately changed since your refactoring of the filestreams.

In the end I need something similar for the memory-mapped file. It should respect that the file was created with FileOptions.DeleteOnClose because then the file will be created nonetheless and there is no need to write the data.

Does this make sense?

**Comment by adamsitnik** [comment]
@msedi thank you for your input! I've analyzed the code and want to share an update.

First of all, I've found a workaround that get can you unblocked:

`MemoryMappedViewAccessor.Dispose` won't call `Flush` if the view is already closed.

https://github.com/dotnet/runtime/blob/bffe34e0c7ff8a05e79d884ed8447426aae17bfb/src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedViewAccessor.cs#L37-L39

We can close the view by using the `SafeMemoryMappedViewHandle` property and calling `Close` for it:

```cs
string path = Path.GetTempFileName();
using FileStream file = new(path, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.None, bufferSize: 0, FileOptions.DeleteOnClose);
using MemoryMappedFile mmf = MemoryMappedFile.CreateFromFile(file, null, 100_000, MemoryMappedFileAccess.Write, HandleInheritability.None, leaveOpen: true);
using MemoryMappedViewAccessor mmva = mmf.CreateViewAccessor(offset: 0, size: 100_000);

// perform some writes etc

// workaround: call Close on the handle first, so MemoryMappedViewAccessor.Dispose does not call Flush
mmva.SafeMemoryMappedViewHandle.Close();
```

Beside that I have verified that making `FileStream.Flush` a nop for such files would not solve the problem, as it's not the `FileStream` that is being flushed, but the view. The view also needs to be closed, so `bool leaveOpen` does not make sense. That is why we need `bool flushOnDispose = true` (I am sure the API review board is going to find a better name). I am going to post a separate comment with the proposal.

**Comment by adamsitnik** [comment]
### Background and motivation

In scenarios where users create temporary files with `FileOptions.DeleteOnClose` and then `MemoryMappedFile` out of it, performing a flush of the view on dispose may be redundant (the file will be deleted very soon, so there is no need to flush to disk first).

### API Proposal

```diff
namespace System.IO.MemoryMappedFiles;

public partial class MemoryMappedFile : IDisposable
{
   public MemoryMappedViewAccessor CreateViewAccessor()
   public MemoryMappedViewAccessor CreateViewAccessor(long offset, long size)
   public MemoryMappedViewAccessor CreateViewAccessor(long offset, long size, MemoryMappedFileAccess access);
+  public MemoryMappedViewAccessor CreateViewAccessor(long offset, long size, MemoryMappedFileAccess access, bool flushOnDispose = true);
}
```

### API Usage

```csharp
// Create a temporary FileStream with FileOptions.DeleteOnClose
string path = Path.GetTempFileName();
using FileStream file = new(path, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.None, bufferSize: 0, FileOptions.DeleteOnClose);

using MemoryMappedFile mmf = MemoryMappedFile.CreateFromFile(file, null, 100_000, MemoryMappedFileAccess.Write, HandleInheritability.None, leaveOpen: true);
using MemoryMappedViewAccessor mmva = mmf.CreateViewAccessor(offset: 0, size: 100_000, MemoryMappedFileAccess.Write,
    flushOnDispose: false); // the new API

// perform some writes etc
```

### Risks

Some users may provide wrong value for non-temporary files and introduce bugs. Providing the default value should lower that risk.

**Comment by adamsitnik** [comment]
Note to future self: I've implemented it to see if it's possible to have this overload with default value (https://github.com/dotnet/runtime/compare/main...adamsitnik:runtime:mmfDontFlush)

**Comment by msedi** [comment]
@adamsitnik: Thanks for the feedback.

> We can close the view by using the SafeMemoryMappedViewHandle property and calling Close for it:

Thanks. That's what I also found wo workaround the problem currently.

Regarding your API suggestion, I think it totally makes sense to do it that way. I currently do not have a better name for the flushOnDispose.

**Comment by terrajobst** [comment]
* Instead of exposing the new parameter, could we handle this automatically?
* If we expose it, should `size` and `access` be defaulted?
    - If so, we should hide the existing second and third overload
    - We should keep the first for simplicity

```C#
namespace System.IO.MemoryMappedFiles;

public partial class MemoryMappedFile
{
    // Existing overloads:
    // public MemoryMappedViewAccessor CreateViewAccessor()
    // public MemoryMappedViewAccessor CreateViewAccessor(long offset, long size)
    // public MemoryMappedViewAccessor CreateViewAccessor(long offset, long size, MemoryMappedFileAccess access);
    public MemoryMappedViewAccessor CreateViewAccessor(long offset, long size, MemoryMappedFileAccess access, bool flushOnDispose = true);
}
```

**Comment by adamsitnik** [comment]
> That's what I also found wo workaround the problem currently.

@msedi Could you please write a benchmark that mimics your scenario and use that workaround to find out how much perf we could gain? Thanks!
